File: aat-android/src/main/java/ch/bailu/aat/gpx/GpxListArray.java
package ch.bailu.aat.gpx;

import ch.bailu.aat.gpx.linked_list.Node;

public class GpxListArray {
    private int index=0;
    private final GpxList list;
    private Node current;


    public GpxListArray(GpxList l) {
        list = l;
        current = list.getPointList().getFirst();
    }

    public GpxPointNode get(int i) {
        setIndex(i);
        return get();
    }

    public void setIndex(int i) {
        while (index < i) {
            current= current.getNext();
            index++;
        }

        while (index > i) {
            current = current.getPrevious();
            index--;
        }
    }

    public GpxPointNode get() {
        return (GpxPointNode) current;
    }
    public int size() {
        return list.getPointList().size();
    }
    public GpxList getList() {
        return list;
    }
    public int getIndex() {
        return index;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/GpxListArray.kt
package ch.bailu.aat_lib.gpx

import ch.bailu.aat_lib.gpx.attributes.GpxAttributesNull
import ch.bailu.aat_lib.gpx.linked_list.Node

/** Index-based random access over a [GpxList]'s linked list of points, caching the current position. */
class GpxListArray(val list: GpxList) {
    var index = 0
        set(i) {
            while (field < i) {
                current = current?.next
                field++
            }
            while (field > i) {
                current = current?.previous
                field--
            }
        }
    private var current: Node?

    init {
        current = list.pointList.first
    }

    operator fun get(i: Int): GpxPointNode {
        index = i
        return get()
    }

    fun get(): GpxPointNode {
        val result = current

        if (result is GpxPointNode) {
            return result
        }
        return GpxPointFirstNode(GpxPoint.NULL, GpxAttributesNull.NULL)
    }

    fun size(): Int {
        return list.pointList.size()
    }
}
